File: v2/src/main/mongo/find.ts
import { EJSON } from 'bson';
import type { MongoClient } from 'mongodb';
import type { FindRequest, FindResult } from '../../shared/query.js';
import { parseFilter, parseProjection, parseSort } from './queryParser.js';

const PAGE_FETCH_EXTRA = 1;

export async function executeFind(
  client: MongoClient,
  req: FindRequest,
): Promise<FindResult> {
  const t0 = performance.now();

  const filter = parseFilter(req.filter);
  const projection = parseProjection(req.projection);
  const sort = parseSort(req.sort);
  const skip = clampNonNegative(req.skip);
  const limit = clampPositive(req.limit, 50);

  let cursor = client.db(req.db).collection(req.collection).find(filter);

  if (Object.keys(projection).length > 0) cursor = cursor.project(projection);
  if (Object.keys(sort).length > 0) cursor = cursor.sort(sort);
  if (skip > 0) cursor = cursor.skip(skip);

  // Fetch one extra row to determine hasMore without exhausting cursor.
  cursor = cursor.limit(limit + PAGE_FETCH_EXTRA);

  if (req.maxTimeMs && req.maxTimeMs > 0) {
    cursor = cursor.maxTimeMS(req.maxTimeMs);
  }

  const fetched = await cursor.toArray();
  const hasMore = fetched.length > limit;
  const rows = (hasMore ? fetched.slice(0, limit) : fetched).map((d) =>
    EJSON.stringify(d, { relaxed: false }),
  );

  return {
    ok: true,
    rows,
    hasMore,
    durationMs: Math.round(performance.now() - t0),
  };
}

function clampNonNegative(n: number | undefined): number {
  if (!Number.isFinite(n)) return 0;
  return Math.max(0, Math.floor(n!));
}

function clampPositive(n: number | undefined, fallback: number): number {
  if (!Number.isFinite(n) || (n ?? 0) <= 0) return fallback;
  return Math.max(1, Math.floor(n!));
}

File: v3/src/main/kotlin/io/mex/mongo/Find.kt
package io.mex.mongo

import com.mongodb.client.MongoClient
import org.bson.Document
import org.bson.json.JsonMode
import org.bson.json.JsonWriterSettings
import java.util.concurrent.TimeUnit

data class FindRequest(
    val connectionId: String,
    val db: String,
    val collection: String,
    val filter: String = "",
    val projection: String = "",
    val sort: String = "",
    val skip: Int = 0,
    val limit: Int = 50,
    val maxTimeMs: Long = 30_000,
)

sealed class FindResult {
    data class Ok(
        val rows: List<String>, // canonical EJSON
        val hasMore: Boolean,
        val durationMs: Long,
    ) : FindResult()
    data class Failed(val error: String, val durationMs: Long) : FindResult()
}

private val EJSON_SETTINGS: JsonWriterSettings = JsonWriterSettings.builder()
    .outputMode(JsonMode.EXTENDED)
    .build()

fun executeFind(client: MongoClient, req: FindRequest): FindResult {
    val t0 = System.nanoTime()
    return runCatching {
        val filter = parseFilter(req.filter)
        val projection = parseProjection(req.projection)
        val sort = parseSort(req.sort)
        val limit = req.limit.coerceAtLeast(1)
        val pageFetch = limit + 1

        var cursor = client.getDatabase(req.db).getCollection(req.collection).find(filter)
        if (!projection.isEmpty) cursor = cursor.projection(projection)
        if (!sort.isEmpty) cursor = cursor.sort(sort)
        if (req.skip > 0) cursor = cursor.skip(req.skip)
        cursor = cursor.limit(pageFetch)
        if (req.maxTimeMs > 0) cursor = cursor.maxTime(req.maxTimeMs, TimeUnit.MILLISECONDS)

        val docs = cursor.into(mutableListOf<Document>())
        val hasMore = docs.size > limit
        val rows = (if (hasMore) docs.dropLast(1) else docs).map { it.toJson(EJSON_SETTINGS) }
        FindResult.Ok(
            rows = rows,
            hasMore = hasMore,
            durationMs = (System.nanoTime() - t0) / 1_000_000,
        )
    }.getOrElse {
        FindResult.Failed(
            error = it.message ?: it::class.java.simpleName,
            durationMs = (System.nanoTime() - t0) / 1_000_000,
        )
    }
}
